cmd/encryption-v1.go

```diff
@@ -412,16 +441,191 @@ func DecryptRequest(client io.Writer, r *http.Request, metadata map[string]strin
 	return DecryptRequestWithSequenceNumber(client, r, 0, metadata)
 }
 
-// getStartOffset - get sequence number, start offset and rlength.
-func getStartOffset(offset, length int64) (seqNumber uint32, startOffset int64, rlength int64) {
-	seqNumber = uint32(offset / (64 * 1024))
-	startOffset = int64(seqNumber) * (64*1024 + 32)
+// DecryptBlocksWriter - decrypts multipart parts, while implementing a io.Writer compatible interface.
+type DecryptBlocksWriter struct {
+	// Original writer where the plain data will be written
+	writer io.Writer
+	// Current decrypter for the current encrypted data block
+	decrypter io.WriteCloser
+	// Start sequence number
+	startSeqNum uint32
+	// Current part index
+	partIndex int
+	// Parts information
+	parts    []objectPartInfo
+	req      *http.Request
+	metadata map[string]string
+
+	partEncRelOffset int64
+
+	// Customer Key
+	customerKeyHeader string
+}
 
-	rlength = (length / (64 * 1024)) * (64*1024 + 32)
-	if length%(64*1024) > 0 {
-		rlength += 64*1024 + 32
+func (w *DecryptBlocksWriter) buildDecrypter(partID int) error {
+	m := make(map[string]string)
+	for k, v := range w.metadata {
+		m[k] = v
 	}
-	return seqNumber, startOffset, rlength
+	// Initialize the first decrypter, new decrypters will be initialized in Write() operation as needed.
+	w.req.Header.Set(SSECustomerKey, w.customerKeyHeader)
+	key, err := ParseSSECustomerRequest(w.req)
+	if err != nil {
+		return err
+	}
+
+	objectEncryptionKey, err := decryptObjectInfo(key, m)
+	if err != nil {
+		return err
+	}
+
+	sha := sha256.New() // derive part  encryption key
+	sha.Write(objectEncryptionKey)
+	sha.Write([]byte("-" + fmt.Sprintf("%d", partID)))
+	partEncryptionKey := sha.Sum(nil)
+
+	delete(m, SSECustomerKey) // make sure we do not save the key by accident
+
+	decrypter, err := newDecryptWriterWithObjectKey(w.writer, partEncryptionKey, w.startSeqNum, m)
+	if err != nil {
+		return err
+	}
+	w.decrypter = decrypter
+	return nil
+}
+
+func (w *DecryptBlocksWriter) Write(p []byte) (int, error) {
+	var err error
+	var n1 int
+	if int64(len(p)) < w.parts[w.partIndex].Size-w.partEncRelOffset {
+		n1, err = w.decrypter.Write(p)
+		if err != nil {
+			return 0, err
+		}
+		w.partEncRelOffset += int64(n1)
+	} else {
+		n1, err = w.decrypter.Write(p[:w.parts[w.partIndex].Size-w.partEncRelOffset])
+		if err != nil {
+			return 0, err
+		}
+
+		// We should now proceed to next part, reset all values appropriately.
+		w.partEncRelOffset = 0
+		w.startSeqNum = 0
+
+		w.partIndex++
+
+		err = w.buildDecrypter(w.partIndex + 1)
+		if err != nil {
+			return 0, err
+		}
+
+		n1, err = w.decrypter.Write(p[n1:])
+		if err != nil {
+			return 0, err
+		}
+
+		w.partEncRelOffset += int64(n1)
+	}
+
+	return len(p), nil
+}
+
+// Close closes the LimitWriter. It behaves like io.Closer.
+func (w *DecryptBlocksWriter) Close() error {
+	if closer, ok := w.writer.(io.Closer); ok {
```

Comment:
You are supposed to close the `decrypter` here - otherwise you miss internally buffered data.
See https://github.com/minio/sio/blob/master/dare.go#L151

Reply:
> You are supposed to close the decrypter here - otherwise you miss internally buffered data.
> See https://github.com/minio/sio/blob/master/dare.go#L151

Done